### PR TITLE
Implementation of Characteristic descriptor discovery

### PR DESCRIPTION
**Pull request by pan-**
This is the implementation of https://github.com/ARMmbed/ble/pull/105

The characteristic discovery process has been slightly modified. To discover characteristic descriptors, it is required to know the start handle and end handle of a characteristic. The discovery process handle this now.

Limitations: 
  - Currently, Characteristics Descriptors 128 bits UUID are not managed properly.

### CODE REVIEW

File: source/btle/btle.cpp [hunk 1]
```diff
@@ -183,6 +183,7 @@ static void btle_handler(ble_evt_t *p_ble_evt)
                     break;
             }
             nRF5xGap::getInstance().processDisconnectionEvent(handle, reason);
+            // TODO: close pending discoveries
```
Comment (pan-) [inline source/btle/btle.cpp]: This is a question. Is it required to close pending discovery manually ? If a discovery is running, does the stack will report the discovery that the discovery has failled (service, characteristic and characteristic descriptors) ?

Comment (rgrover) [inline source/btle/btle.cpp]: I believe we should invoke the serviceDiscoveryTermination callback before invoking disconnectionCallback. Not sure yet how to report a failure to service discovery.

Comment (pan-) [inline source/btle/btle.cpp]: I have made the changes. I have added a new function in `nRF5xServiceDiscovery` : `void terminate(Gap::Handle_t connectionHandle)`. It is an overload of terminate, if the handle passed in parameters match the handle of the current discovery, it terminate the discovery otherwise, the discovery continue to run.

File: source/btle/btle.cpp [hunk 2]
```diff
@@ -182,6 +186,12 @@ static void btle_handler(ble_evt_t *p_ble_evt)
                     reason = static_cast<Gap::DisconnectionReason_t>(p_ble_evt->evt.gap_evt.params.disconnected.reason);
                     break;
             }
+
+            // Close all pending discoveries for this connection
+            nRF5xGattClient& gattClient = nRF5xGattClient::getInstance();
+            gattClient.characteristicDescriptorDiscoverer().terminate(handle, BLE_ERROR_INVALID_STATE);
+            gattClient.discovery().terminate(handle);
```
Comment (rgrover) [inline source/btle/btle.cpp]: thanks for this improvement. :)

File: source/btle/btle_discovery.cpp [hunk 3]
```diff
@@ -44,7 +47,7 @@ void bleGattcEventHandler(const ble_evt_t *p_ble_evt)
 
                 case BLE_GATT_STATUS_ATTERR_ATTRIBUTE_NOT_FOUND:
                 default:
-                    sdSingleton.terminateCharacteristicDiscovery();
+                    sdSingleton.terminateCharacteristicDiscovery(BLE_ERROR_NONE);
```
Comment (rgrover) [inline source/btle/btle_discovery.cpp]: you're combining several different changes into a single pull request. This makes it difficult for the reviewer to manage complexity. Please try to keep pull requests focussed on a single topic at a time. 

If you wanted to improve the API for `terminateCharacteristicDiscovery`, you could have done that separately as an independent pull request.

File: source/nRF5xCharacteristicDescriptorDiscoverer.h [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+#include "ble/Gap.h"
+#include "ble/DiscoveredCharacteristic.h"
+#include "ble/CharacteristicDescriptorDiscovery.h"
+#include "ble/GattClient.h"
+#include "ble_gattc.h"
+
+/**
+ * @brief Manage the discovery of Characteristic descriptors 
+ * @details is a bridge beetween BLE API and nordic stack regarding Characteristic
+ * Descriptor discovery. The BLE API can launch, monitorate and ask for termination
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.h]: `monitorate`?

File: source/nRF5xCharacteristicDescriptorDiscoverer.h [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+ * of a discovery. The nordic stack will provide new descriptors and indicate when
+ * the discovery is done
+ */
+class nRF5xCharacteristicDescriptorDiscoverer
+{
+    typedef CharacteristicDescriptorDiscovery::DiscoveryCallback_t DiscoveryCallback_t;
+    typedef CharacteristicDescriptorDiscovery::TerminationCallback_t TerminationCallback_t;
+
+public:
+    nRF5xCharacteristicDescriptorDiscoverer(size_t concurrentConnectionsCount = 3);
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.h]: docs

File: source/nRF5xCharacteristicDescriptorDiscoverer.h [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**
+     * @brief indicate if a characteristic descriptor discovery is active for a 
+     * given DiscoveredCharacteristic
+     * @note: this will be called by BLE API side
+     */
+    bool isActive(const DiscoveredCharacteristic& characteristic) const;
+
+    /**
+     * @brief reauest the termination of characteristic descriptor discovery
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.h]: `reauest`

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+namespace { 
+    void emptyDiscoveryCallback(const CharacteristicDescriptorDiscovery::DiscoveryCallbackParams_t*) { }
+    void emptyTerminationCallback(const CharacteristicDescriptorDiscovery::TerminationCallbackParams_t*) { }
+}
+
+nRF5xCharacteristicDescriptorDiscoverer::nRF5xCharacteristicDescriptorDiscoverer(size_t concurrentConnectionsCount) : 
+    maximumConcurrentConnectionsCount(concurrentConnectionsCount), 
+    discoveryRunning(new Discovery[concurrentConnectionsCount]) {
+
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: delete empty line

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+#include "ble/DiscoveredCharacteristicDescriptor.h"
+
+namespace { 
+    void emptyDiscoveryCallback(const CharacteristicDescriptorDiscovery::DiscoveryCallbackParams_t*) { }
+    void emptyTerminationCallback(const CharacteristicDescriptorDiscovery::TerminationCallbackParams_t*) { }
+}
+
+nRF5xCharacteristicDescriptorDiscoverer::nRF5xCharacteristicDescriptorDiscoverer(size_t concurrentConnectionsCount) : 
+    maximumConcurrentConnectionsCount(concurrentConnectionsCount), 
+    discoveryRunning(new Discovery[concurrentConnectionsCount]) {
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: can we avoid allocation here?

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+ble_error_t nRF5xCharacteristicDescriptorDiscoverer::launch(
+    const DiscoveredCharacteristic& characteristic, 
+    const CharacteristicDescriptorDiscovery::DiscoveryCallback_t& discoveryCallback,
+    const CharacteristicDescriptorDiscovery::TerminationCallback_t& terminationCallback
+) {
+    Gap::Handle_t connHandle = characteristic.getConnectionHandle();
+    // it is ok to deduce that the start handle for descriptors is after 
+    // the characteristic declaration and the characteristic value declaration
+    // see BLUETOOTH SPECIFICATION Version 4.2 [Vol 3, Part G] (3.3)
+    Gap::Handle_t descriptorStartHandle = characteristic.getDeclHandle() + 2;
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: if we don't do the `+ 2`, it should still work, shouldn't it? The char. descriptor and value attribute would not be returned by the stack as descriptors.

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    Gap::Handle_t descriptorEndHandle = characteristic.getLastHandle();
+
+    // check if their is any descriptor to discover
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: `their` should be `there`

File: source/btle/btle_discovery.cpp (hunk 3 above)
Comment (pan-) [inline source/btle/btle_discovery.cpp]: Discovery of descriptor require the knowledge of the last ATT handle of a characteristic definition. This information wasn't captured before during characteristic discovery, that's why I was forced to change some things in characteristic discovery. 

Knowing that the characteristic discovery end gracefully allow me to set the last handle of the last characteristic discovered to the last handle of its service.

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp (hunk 8 above)
Comment (pan-) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: Yes, is it ok if I made nRF5xCharacteristicDescriptorDiscoverer a template class with the count of concurrent connections as a non type template parameter ?

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp (hunk 9 above)
Comment (pan-) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: According to the specification, no:

The Attribute Protocol Find Information Request shall be used with the Starting
Handle set to the handle of the specified characteristic value + 1 and the
Ending Handle set to the ending handle of the specified characteristic.

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    return findRunningDiscovery(characteristic) != NULL;
+}
+
+void nRF5xCharacteristicDescriptorDiscoverer::requestTerminate(const DiscoveredCharacteristic& characteristic) {
+    Discovery* discovery = findRunningDiscovery(characteristic);
+    if(discovery) { 
+        discovery->onDiscovery = emptyDiscoveryCallback;
+        // call terminate anyway
+        discovery->terminate(BLE_ERROR_NONE);
+        discovery->onTerminate = emptyTerminationCallback;
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: is this reset needed?

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    Discovery tmp = *discovery;
+    *discovery = Discovery();
+    tmp.terminate(err);
+}
+
+nRF5xCharacteristicDescriptorDiscoverer::Discovery* 
+nRF5xCharacteristicDescriptorDiscoverer::findRunningDiscovery(const DiscoveredCharacteristic& characteristic) {
+    for(size_t i = 0; i < maximumConcurrentConnectionsCount; ++i) {
```
Comment (rgrover) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: could this simply forward the call to `(const_cast<const nRF5xCharacteristicDescriptorDiscoverer *>(this))->findRunningDiscovery()`?

File: source/nRF5xServiceDiscovery.cpp [hunk 13]
```diff
@@ -114,38 +123,63 @@ nRF5xServiceDiscovery::setupDiscoveredCharacteristics(const ble_gattc_evt_char_d
 void
 nRF5xServiceDiscovery::progressCharacteristicDiscovery(void)
 {
-    /* Iterate through the previously discovered characteristics cached in characteristics[]. */
-    while ((state == CHARACTERISTIC_DISCOVERY_ACTIVE) && (characteristicIndex < numCharacteristics)) {
+    if(state != CHARACTERISTIC_DISCOVERY_ACTIVE) {
+        return;
+    }
+
+    if(remainingCharacteristic != nRF5xDiscoveredCharacteristic() && numCharacteristics > 0) {
```
Comment (rgrover) [inline source/nRF5xServiceDiscovery.cpp]: it would be nice to enclose multiple boolean sub-expressions into parenthesis. 
`((remainingCharacteristic != nRF5xDiscoveredCharacteristic()) && (numCharacteristics > 0))`

Comment (rgrover) [inline source/nRF5xServiceDiscovery.cpp]: why are you changing this?
This seems to be off topic.

File: source/nRF5xServiceDiscovery.h [hunk 14]
```diff
@@ -300,6 +321,8 @@ class nRF5xServiceDiscovery : public ServiceDiscovery
     CharUUIDDiscoveryQueue      charUUIDDiscoveryQueue;
 
     TerminationCallback_t       onTerminationCallback;
+
+    nRF5xDiscoveredCharacteristic remainingCharacteristic;
```
Comment (rgrover) [inline source/nRF5xServiceDiscovery.h]: what is this `remainingCharacteristic`? It seems to be key to the new implementation. I would have liked to find a comment describing its function.

File: source/nRF5xServiceDiscovery.cpp [hunk 15]
```diff
@@ -114,38 +123,63 @@ nRF5xServiceDiscovery::setupDiscoveredCharacteristics(const ble_gattc_evt_char_d
...
-                .end_handle   = endHandle
-            };
-            if (sd_ble_gattc_characteristics_discover(connHandle, &handleRange) != NRF_SUCCESS) {
-                terminateCharacteristicDiscovery();
+    for(uint8_t i = 0; i < numCharacteristics; ++i) {
+        if(state != CHARACTERISTIC_DISCOVERY_ACTIVE) {
+            return;
+        }
+
+        if(i == numCharacteristics - 1) {
```
Comment (rgrover) [inline source/nRF5xServiceDiscovery.cpp]: `(i == (numCharacteristics - 1))`

File: source/nRF5xServiceDiscovery.cpp [hunk 16]
```diff
@@ -114,38 +123,63 @@ nRF5xServiceDiscovery::setupDiscoveredCharacteristics(const ble_gattc_evt_char_d
...
-        } else {
-            terminateCharacteristicDiscovery();
         }
     }
+
+    if(state != CHARACTERISTIC_DISCOVERY_ACTIVE) {
+        return;
+    }
+
+
```
Comment (rgrover) [inline source/nRF5xServiceDiscovery.cpp]: extra blank line

**Comment by rgrover** [comment]
I've finished my review for the moment. I'll wait for your updates.

File: source/nRF5xCharacteristicDescriptorDiscoverer.cpp (hunk 11 above)
Comment (pan-) [inline source/nRF5xCharacteristicDescriptorDiscoverer.cpp]: Yes, this is the request to terminate the discovery process, it does not mean that anything is done at this point. The termination will occur from the stack side (see the function `terminate` or `process`) and will call the terminate member of the discovery. 

The `onTerminate` member shall better be something valid, replacing the user onTerminate function by an empty one seems the right choice for me. 

Otherwise it is possible to add special member flag indicating if the discovery is terminated from the user side and test if this flag is set every time we try to call terminate but I have the feeling that it makes the code less elegant, less efficient since the condition should always be tested and less understandable since there will be flags test all over the place.

**Comment by pan-** [comment]
@rgrover, @andresag01 Could you review these changes ?
